sound/soc/sof/topology.c

```diff
@@ -3515,6 +3515,45 @@ static int sof_set_pipe_widget(struct snd_sof_dev *sdev, struct snd_sof_widget *
 	return 0;
 }
 
+/*
+ * Function to get the initial value of SOF kcontrols. The value will be stored in
+ * the private data in struct snd_sof_control. This is particularly useful for binary kcontrols
+ * associated with static pipeline widgets to get the data size from the DSP.
+ */
+static int snd_sof_cache_kcontrol_val(struct snd_soc_component *scomp)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_control *scontrol;
+	int ipc_cmd, ctrl_type;
+	int ret = 0;
+
+	list_for_each_entry(scontrol, &sdev->kcontrol_list, list) {
+		struct snd_sof_widget *swidget;
+
+		swidget = snd_sof_find_swidget_by_id(scomp, scontrol->comp_id);
+		if (!swidget) {
+			dev_warn(scomp->dev, "No widget found with ID %d for control\n",
+				 scontrol->comp_id);
+			continue;
+		}
+
+		/*
+		 * Dynamic pipelines are not set up at this point. So only perform get for
+		 * static pipeline widgets.
+		 */
+		if (swidget->dynamic_pipeline_widget)
+			continue;
+
+		/* get kcontrol values from DSP */
+		ret = snd_sof_ipc_set_get_comp_data(scontrol, false);
+		if (ret < 0)
+			dev_warn(scomp->dev, "Failed kcontrol get for control in widget %s\n",
+				 swidget->widget->name);
```

Comment:
Print the control name as well?

We only return error code if the data get fails for the _last_ control? If the last succeeds in the list, but all prior ones fail then we return a happy 0..